Feature schedule posts

Rajat-Dabade opened the pull request:
#### Summary
Main PR for Feature Schedule Posts

#### Ticket Link
<!--
If this pull request addresses a Help Wanted ticket or fixes a reported issue, please link the relevant GitHub issue, e.g.

  Fixes https://github.com/mattermost/mattermost-mobile/issues/XXXXX

Otherwise, link the JIRA ticket.
-->

#### Checklist
<!--
Place an '[x]' (no spaces) in all applicable fields. Please remove unrelated fields.
-->
- [] Added or updated unit tests (required for all new features)
- [] Has UI changes
- [ ] Includes text changes and localization file updates
- [ ] Have tested against the 5 core themes to ensure consistency between them.
- [ ] Have run E2E tests by adding label `E2E iOS tests for PR`.

#### Device Information
This PR was tested on: <!-- Device name(s), OS version(s) -->

#### Screenshots
<!--
If the PR includes UI changes, include screenshots/GIFs/Videos (for both iOS and Android if possible).
-->

#### Release Note
<!--
Add a release note for each of the following conditions:

* New features and improvements, including behavioural changes, UI changes
* Bug fixes and fixes of previous known issues
* Deprecation warnings, breaking changes, or compatibility notes

If no release notes are required write NONE. Use past-tense. Newlines are stripped.

Example:

```release-note
Added a new config setting ServiceSettings.FooBar. Added a new column Foo to the Users table.
```

```release-note
NONE
```
-->

```release-note
Added a feature to support Schedule Post in mobile
```

Rajat-Dabade commented:
/update-branch

Rajat-Dabade commented:
/update-branch

mattermost-build commented:
Error trying to update the PR.
Please do it manually.

Rajat-Dabade commented:
/update-branch

harshilsharma63 commented on app/client/rest/base.ts:
```diff
@@ -171,6 +171,14 @@ export default class ClientBase extends ClientTracking {
         return `${this.urlVersion}/redirect_location`;
     }
 
+    getScheduledPostActionsRoute() {
```
This is the same as `getScheduledPostRoute`, so can we remove this?

Rajat-Dabade replied:
Sure

harshilsharma63 commented on app/screens/global_drafts/components/tabbed_contents/draftTabsHeader.tsx:
```diff
@@ -0,0 +1,142 @@
+
+    const draftCountBadge = useMemo(() => {
+        // eslint-disable-next-line no-warning-comments
+        // TODO: when integrating with database, handle here the case of draft count being 0, and return undefined to show no badge
```
Can we implement this todo? If draft or scheduled post is 0, there should be no badge in the tab. Currently, I believe it will show a badge with count 0.

Rajat-Dabade replied:
Done.

harshilsharma63 commented on app/screens/global_drafts/components/tabbed_contents/draftTabsHeader.tsx:
```diff
@@ -0,0 +1,142 @@
+
+    const scheduledPostCountBadge = useMemo(() => {
+        // eslint-disable-next-line no-warning-comments
+        // TODO: when integrating with database, handle here the case of scheduled post count being 0, and return undefined to show no badge
```
Same here.

Rajat-Dabade replied:
Done.

harshilsharma63 commented on ios/Gekidou/Package.resolved:
```diff
@@ -1,13 +1,94 @@
 {
```
This is probably an unintended change. If so, can we restore this from whats on `master`?

Rajat-Dabade replied:
Sure

harshilsharma63 reviewed (changes requested):
Just a couple of minor things.

Rajat-Dabade commented on app/screens/home/channel_list/additional_tablet_view/additional_tablet_view.tsx:
```diff
@@ -28,10 +31,17 @@ const ComponentsList: Record<string, React.ComponentType<any>> = {
 const channelScreen: SelectedView = {id: Screens.CHANNEL, Component: Channel};
-const globalScreen: SelectedView = {id: Screens.GLOBAL_THREADS, Component: GlobalThreads};
+const globalThreadsScreen: SelectedView = {id: Screens.GLOBAL_THREADS, Component: GlobalThreads};
+const globalDraftsScreen: SelectedView = {id: Screens.GLOBAL_THREADS, Component: GlobalDraftsAndScheduledPosts};
```
Should this be `{id: Screens.GLOBAL_DRAFTS}`?

enahum replied:
Yes

Rajat-Dabade replied:
Made changes.

Rajat-Dabade commented:
/update-branch

larkox commented on app/queries/servers/team.test.ts:
```diff
@@ -1313,4 +1324,56 @@ describe('Team Queries', () => {
+                prepareRecordsOnly: false,
+            });
+
+            observeTeamLastChannelId(database, teamId).subscribe((lastChannelId) => {
+                expect(lastChannelId).toBe(channelId);
+            });
```
Almost sure the test is finishing before the `expect` is called, so changing `channelID` for anything (like 'foo') will still make the test pass. I think I had some example of how to properly test this in this PR: https://github.com/mattermost/mattermost-mobile/pull/8624/files

larkox commented on app/queries/servers/team.test.ts:
```diff
@@ -1313,4 +1324,56 @@ describe('Team Queries', () => {
+                prepareRecordsOnly: false,
+            });
+
+            const subscription = observeTeamLastChannelId(database, teamId).subscribe((lastChannelId) => {
```
Same as my previous comment.

larkox commented on app/screens/channel/channel.tsx:
```diff
@@ -135,6 +136,9 @@ const Channel = ({
                             nativeID={channelId}
                         />
                     </View>
+                    <ScheduledPostIndicator
+                        channelId={channelId}
+                    />
```
0/5 No tests for this new behavior

larkox commented on app/screens/custom_status_clear_after/components/date_time_selector.tsx:
```diff
@@ -44,14 +43,14 @@ const getStyleSheet = makeStyleSheetFromTheme((theme: Theme) => {
     const [date, setDate] = useState<Moment>(minimumDate);
-    const [mode, setMode] = useState<AndroidMode>('date');
-    const [show, setShow] = useState<boolean>(false);
+    const [mode, setMode] = useState<AndroidMode>(showInitially || 'date');
+    const [show, setShow] = useState<boolean>(Boolean(showInitially));
```
No tests for this new behavior.

larkox commented on assets/base/images/Scheduled_Message.png:
@asaadmahmood  Is this still the correct image to use? With the branding changes this one looks a bit off, but I may be mistaken.

larkox commented on app/utils/snack_bar/index.ts:
```diff
@@ -54,3 +63,12 @@ export const showThreadFollowingSnackbar = (following: boolean, onAction: () =>
     });
 };
+
+export const showScheduledPostCreationErrorSnackbar = (errorMessage: string) => {
```
0/5 too simple of a function, but there are no tests for it.

larkox reviewed (changes requested):
98 out of 146 files reviewed. I am missing still mostly the screens folder and the utils folder.

Rajat-Dabade commented on app/actions/app/global.ts:
```diff
@@ -44,6 +44,14 @@ export const storeDraftsTutorial = async () => {
+    return storeGlobal(Tutorial.SCHEDULED_POST, 'true', false);
+};
+
+export const storeScheduledPostsTutorial = async () => {
```
Changed the name to  `storeScheduledPostsListTutorial` and added a test.

Rajat-Dabade commented on app/actions/local/draft.ts:
```diff
@@ -6,17 +6,61 @@ import {DeviceEventEmitter, Image} from 'react-native';
+        }
+        const params: goToScreenParams = {};
+
+        // explicitly checking for undefined because 0 is a valid value
+        // that shouldn't be missed in this case.
+        if (initialTab !== undefined) {
+            params.initialTab = initialTab;
+        }
```
Yup, that's true.

Rajat-Dabade commented on app/actions/local/draft.test.ts:
```diff
@@ -244,6 +261,75 @@ describe('updateDraftPriority', () => {
+        expect(emitSpy).not.toHaveBeenCalled();
+    });
+
+    it('should call NOT goToScreen on non-tablet', async () => {
```
Changed it.

Rajat-Dabade commented on app/actions/local/draft.test.ts:
```diff
@@ -244,6 +261,75 @@ describe('updateDraftPriority', () => {
+        expect(emitSpy).not.toHaveBeenCalled();
+    });
+
+    it('should pass initialTab param when provided', async () => {
```
Added test for non tablet case

Rajat-Dabade commented on app/actions/local/draft.ts:
```diff
@@ -6,17 +6,61 @@ import {DeviceEventEmitter, Image} from 'react-native';
+        const history = await addChannelToTeamHistory(operator, teamIdToUse, Screens.GLOBAL_DRAFTS, true);
+        models.push(...history);
+
+        if (!prepareRecordsOnly) {
+            await operator.batchRecords(models, 'switchToGlobalThreads');
+        }
```
One way to test this is to spy on the batchRecords and check whether it has been called or not. I have added that test.

Rajat-Dabade commented on app/actions/local/scheduled_post.test.ts:
```diff
@@ -0,0 +1,155 @@
+
+    it('handleScheduledPosts - should create scheduled post', async () => {
+        const {models} = await handleScheduledPosts(serverUrl, ActionType.SCHEDULED_POSTS.CREATE_OR_UPDATED_SCHEDULED_POST, scheduledPosts);
+        expect(models?.length).toBe(2);
```
`id` of the scheduledPost can be tested. I will add it.

Rajat-Dabade commented on app/actions/local/scheduled_post.test.ts:
```diff
@@ -0,0 +1,155 @@
+        const spyHandledScheduledPosts = jest.spyOn(operator, 'handleScheduledPosts');
+        await handleScheduledPosts(serverUrl, ActionType.SCHEDULED_POSTS.CREATE_OR_UPDATED_SCHEDULED_POST, scheduledPosts);
+        expect(spyHandledScheduledPosts).toHaveBeenCalledTimes(1);
+        expect(spyHandledScheduledPosts).toHaveBeenCalledWith({
```
Added the test for it.

Rajat-Dabade commented on app/actions/local/scheduled_post.test.ts:
```diff
@@ -0,0 +1,155 @@
+        expect(models![0].message).toEqual(updatedScheduledPost.message);
+    });
+
+    it('handleScheduledPosts - should delete scheduled post', async () => {
+        await operator.handleScheduledPosts({
+            actionType: ActionType.SCHEDULED_POSTS.CREATE_OR_UPDATED_SCHEDULED_POST,
+            scheduledPosts,
+            prepareRecordsOnly: false,
+        });
+
+        const {models} = await handleScheduledPosts(
+            serverUrl,
+            ActionType.SCHEDULED_POSTS.DELETE_SCHEDULED_POST,
+            [scheduledPosts[0]],
+            false,
+        );
+
+        expect(models![0].id).toEqual(scheduledPosts[0].id);
+    });
```
I have added 

```
expect(spyBatchRecords).toHaveBeenCalledWith(models, '_deleteScheduledPostByIds');
```

This will prove that the batch with the description `_deleteScheduledPostByIds` is been called.

Rajat-Dabade commented on app/actions/local/scheduled_post.test.ts:
```diff
@@ -0,0 +1,155 @@
+            false,
+        );
+
+        expect(models![0].message).toEqual(updatedScheduledPost.message);
```
Added

Rajat-Dabade commented on app/actions/local/scheduled_post.test.ts:
```diff
@@ -0,0 +1,155 @@
+
+        expect(result.error).toBeDefined();
+        expect(result.error.message).toBe('foo database not found');
+    });
```
Added

Rajat-Dabade commented on app/actions/local/scheduled_post.ts:
```diff
@@ -0,0 +1,34 @@
+
+import type {ScheduledPostErrorCode} from '@utils/scheduled_post';
+
+export async function handleScheduledPosts(serverUrl: string, actionType: string, scheduledPosts: ScheduledPost[], prepareRecordsOnly = false): Promise<{models?: ScheduledPostModel[]; error?: any}> {
```
Changed it to `unknown`.

Rajat-Dabade commented on app/actions/remote/entry/common.ts:
```diff
@@ -305,6 +306,10 @@ export async function restDeferredAppEntryActions(
     // Fetch groups for current user
     fetchGroupsForMember(serverUrl, currentUserId, false, requestLabel);
+
+    if (initialTeamId) {
```
It's added under the test `should execute deferred actions correctly`.

Rajat-Dabade commented on app/actions/remote/scheduled_post.test.ts:
```diff
@@ -0,0 +1,251 @@
+    await DatabaseManager.destroyServerDatabase(serverUrl);
+});
+
+describe('scheduled_post', () => {
```
Done.

Rajat-Dabade commented on app/actions/remote/scheduled_post.test.ts:
```diff
@@ -0,0 +1,251 @@
+});
+
+describe('scheduled_post', () => {
+    it('createScheduledPost - handle not found database', async () => {
```
removed it.

Rajat-Dabade commented on app/actions/remote/scheduled_post.ts:
```diff
@@ -0,0 +1,127 @@
+}
+
+export async function fetchScheduledPosts(serverUrl: string, teamId: string, includeDirectChannels = false, groupLabel?: RequestGroupLabel) {
+    const operator = DatabaseManager.serverDatabases[serverUrl]?.operator;
+    const database = DatabaseManager.serverDatabases[serverUrl]?.database;
```
Made changes. Done.

Rajat-Dabade commented on app/actions/remote/scheduled_post.test.ts:
```diff
@@ -0,0 +1,251 @@
+
+    it('fetch Schedule post - handle scheduled post disabled', async () => {
+        const result = await fetchScheduledPosts(serverUrl, 'bar');
+        expect(result.scheduledPosts).toEqual([]);
```
Sure.

Rajat-Dabade commented on app/actions/websocket/scheduled_post.test.ts:
```diff
@@ -0,0 +1,110 @@
+});
+
+describe('handleCreateOrUpdateSchedulePost', () => {
+    it('handleCreateOrUpdateScheduledPost - handle empty payload', async () => {
```
Yup, updated it.

Rajat-Dabade commented on app/components/draft_scheduled_post/draft_scheduled_post.tsx:
```diff
@@ -0,0 +1,151 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
```
Created a jira ticket for it: will do it later https://mattermost.atlassian.net/browse/MM-62286

Rajat-Dabade commented on app/components/draft_scheduled_post/draft_scheduled_post.tsx:
```diff
@@ -0,0 +1,151 @@
+                        testID='draft_post.channel_info'
+                        updateAt={post.updateAt}
+                        draftType={draftType}
+                        postScheduledAt={(post as ScheduledPostModel).scheduledAt}
+                        scheduledPostErrorCode={(post as ScheduledPostModel).errorCode}
```
Makes sense, updated it.

larkox commented on app/screens/scheduled_post_options/index.ts:
```diff
@@ -0,0 +1,20 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
```
0/5 really simple, but no tests for this file.

Rajat-Dabade replied:
Added to the list, can be taken up later.

larkox commented on app/screens/thread/thread.tsx:
```diff
@@ -126,6 +127,9 @@ const Thread = ({
                             rootPost={rootPost!}
                         />
                     </View>
+                    <ScheduledPostIndicator
```
0/5 should we add test for this?

Rajat-Dabade replied:
No initial test for this component. Can be taken up later.

Rajat-Dabade replied:
https://mattermost.atlassian.net/browse/MM-62286

larkox commented:
@Rajat-Dabade Another thing I don't see in this feature branch is any e2e test.

Rajat-Dabade commented on app/components/draft_scheduled_post_header/draft_scheduled_post_header.tsx:
```diff
@@ -135,25 +171,44 @@ const DraftPostHeader: React.FC<Props> = ({
+                                size={12}
+                                color={theme.buttonColor}
+                            />
+                            <Text style={style.errorText}>{getErrorStringFromCode(intl, scheduledPostErrorCode as ScheduledPostErrorCode)}</Text>
```
This will be needed as of now, once the errorCode type is fixed: https://mattermost.atlassian.net/browse/MM-63257, this will be removed.

Rajat-Dabade commented:
/update-branch

Rajat-Dabade commented on app/components/drafts_buttton/index.ts:
```diff
@@ -0,0 +1,27 @@
+
+    return {
+        currentChannelId: observeCurrentChannelId(database),
+        lastChannelId: currentTeamId.pipe(
+            switchMap(
+                (teamId) => observeTeamLastChannelId(database, teamId),
+            ),
+        ),
```
Yup, that can be done. Made the changes.

Rajat-Dabade commented on app/components/drafts_buttton/drafts_button.test.tsx:
```diff
@@ -0,0 +1,125 @@
+            <DraftsButton {...props}/>,
+        );
+
+        expect(getByTestId('channel_list.drafts.count')).toHaveTextContent('5');
```
Okay

Rajat-Dabade commented on app/components/formatted_time/index.tsx:
```diff
@@ -15,27 +15,28 @@ type FormattedTimeProps = TextProps & {
     value: number | string | Date;
 }
 
+export function getFormattedTime(isMilitaryTime: boolean, timezone: UserTimezone | string, value: number | string | Date) {
```
Sure, made the changes and added tests.

Rajat-Dabade commented on app/components/post_draft/draft_input/draft_input.tsx:
```diff
@@ -228,10 +262,20 @@ export default function DraftInput({
     );
 }
+
+const enhanced = withObservables([], ({database}: WithDatabaseArgs) => {
+    const scheduledPostsEnabled = observeConfigBooleanValue(database, 'ScheduledPosts');
+    return {
+        scheduledPostsEnabled,
+    };
+});
```
Okay

Rajat-Dabade commented on app/components/post_draft/draft_input/draft_input.test.tsx:
```diff
@@ -0,0 +1,200 @@
+            expect(baseProps.sendMessage).toHaveBeenCalledWith(undefined);
+        });
+
+        it('opens scheduled post options on long press and verify action', async () => {
```
Done.

Rajat-Dabade commented on app/components/post_draft/send_button/index.tsx:
```diff
@@ -0,0 +1,18 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
```
https://mattermost.atlassian.net/browse/MM-62286, will add the test for this component later.

Rajat-Dabade commented on app/components/scheduled_post_indicator_with_datetime/__tests__/scheduled_post_indicator_with_datetime.test.tsx:
```diff
@@ -0,0 +1,73 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
```
Done.

Rajat-Dabade commented on app/components/scheduled_post_indicator/index.ts:
```diff
@@ -0,0 +1,45 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
```
https://mattermost.atlassian.net/browse/MM-62286

enahum commented on app/components/drafts_buttton/index.ts:
```diff
@@ -0,0 +1,31 @@
+
+import type {WithDatabaseArgs} from '@typings/database/database';
+
+const enhanced = withObservables([], ({database, shouldHighlightActive}: {shouldHighlightActive: boolean} & WithDatabaseArgs) => {
```
With the changes made needs be
```suggestion
const enhanced = withObservables(['shouldHighlightActive'], ({database, shouldHighlightActive}: {shouldHighlightActive: boolean} & WithDatabaseArgs) => {
```

Rajat-Dabade commented on app/components/scheduled_post_indicator_with_datetime/index.tsx:
```diff
@@ -0,0 +1,23 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
```
https://mattermost.atlassian.net/browse/MM-62286

Rajat-Dabade commented on app/components/threads_button/index.ts:
```diff
@@ -16,6 +17,11 @@ const enhanced = withObservables([], ({database}: WithDatabaseArgs) => {
 
     return {
         currentChannelId: observeCurrentChannelId(database),
+        lastChannelId: currentTeamId.pipe(
```
No initial written test for this, can be taken up later.

enahum reviewed (changes requested):
Some comments I was about to make without fully reviewing.. sorry!

enahum commented on app/screens/home/channel_list/additional_tablet_view/index.ts:
```diff
@@ -2,23 +2,45 @@
+    );
+
+    const intialView = combineLatest([currentChannelId, lastChannelId, isCRTEnabled]).pipe(
+        first(),
```
Reason being that after the initial view, we don't really want to continue listening for updates

Rajat-Dabade commented on app/actions/websocket/scheduled_post.test.ts:
```diff
@@ -0,0 +1,110 @@
+});
+
+describe('handleDeleteScheduledPost', () => {
+    it('handleDeleteScheduledPost - handle empty payload', async () => {
```
Updated.

Rajat-Dabade commented on app/database/operator/server_data_operator/comparators/scheduled_post.ts:
```diff
@@ -0,0 +1,8 @@
+
+import type ScheduledPostModel from '@typings/database/models/servers/scheduled_post';
+
+export function shouldUpdateScheduledPostRecord(e: ScheduledPostModel, n: ScheduledPost) {
```
Added

Rajat-Dabade commented on app/database/operator/server_data_operator/transformers/post.test.ts:
```diff
@@ -123,4 +124,33 @@ describe('***  POST Prepare Records Test ***', () => {
         expect(preparedRecords!.collection.table).toBe('PostsInChannel');
     });
+
+    it('=> transformScheduledPostRecord: should return an array of type ScheduledPost', async () => {
```
Changed it to `should create a ScheduledPost record from raw data`, 

FYI: I removed the prefix `transformScheduledPostRecord` as well, but the other test in the file has it.

Rajat-Dabade commented on app/database/operator/server_data_operator/transformers/post.test.ts:
```diff
@@ -123,4 +124,33 @@ describe('***  POST Prepare Records Test ***', () => {
         expect(preparedRecords!.collection.table).toBe('PostsInChannel');
     });
+
+    it('=> transformScheduledPostRecord: should return an array of type ScheduledPost', async () => {
```
Added

Rajat-Dabade commented on app/queries/servers/scheduled_post.ts:
```diff
@@ -0,0 +1,56 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
```
Added to the list: https://mattermost.atlassian.net/browse/MM-62286

Rajat-Dabade commented on app/screens/draft_scheduled_post_options/index.tsx:
```diff
@@ -10,19 +10,24 @@ import {Screens} from '@constants';
 import {useTheme} from '@context/theme';
```
Added test.

Rajat-Dabade commented on app/screens/draft_scheduled_post_options/send_draft.tsx:
```diff
@@ -86,8 +94,34 @@ const SendDraft: React.FC<Props> = ({
+                        defaultMessage: 'Post has been create successfully but failed to delete. Please delete the scheduled post manually from the scheduled post list.',
+                    }),
+                    [{
+                        text: intl.formatMessage({id: 'mobile.post.cancel', defaultMessage: 'Cancel'}),
```
@asaadmahmood ⬆️

Rajat-Dabade commented on app/screens/draft_scheduled_post_options/send_draft.test.tsx:
```diff
@@ -0,0 +1,227 @@
+    });
+
+    it('should render the component', () => {
+        const props = {
+            channelId: 'channel_id',
+            channelName: 'channel_name',
+            rootId: '',
+            channelType: General.OPEN_CHANNEL,
+            bottomSheetId: Screens.DRAFT_SCHEDULED_POST_OPTIONS,
+            currentUserId: 'current_user_id',
+            maxMessageLength: 4000,
+            useChannelMentions: true,
+            userIsOutOfOffice: false,
+            customEmojis: [],
+            value: 'value',
+            files: [],
+            postPriority: '' as unknown as PostPriority,
+            persistentNotificationInterval: 0,
+            persistentNotificationMaxRecipients: 0,
+            draftReceiverUserName: undefined,
+            draftType: DRAFT_TYPE_DRAFT,
+        };
```
Done.

Rajat-Dabade commented on app/screens/draft_scheduled_post_options/send_draft.test.tsx:
```diff
@@ -0,0 +1,227 @@
+            customEmojis: [],
+            value: 'value',
+            files: [],
+            postPriority: '' as unknown as PostPriority,
```
True, updated it.

Rajat-Dabade commented on app/screens/draft_scheduled_post_options/send_draft.test.tsx:
```diff
@@ -0,0 +1,227 @@
+        expect(getByText('Send draft')).toBeTruthy();
+    });
+
+    it('should call dismissBottomSheet after sending the draft', () => {
```
Yup, added it.

Rajat-Dabade commented on app/screens/draft_scheduled_post_options/send_draft.test.tsx:
```diff
@@ -0,0 +1,227 @@
+        const {getByText} = wrapper;
+        expect(getByText('Send')).toBeTruthy();
+    });
+
```
Done. Added.

Rajat-Dabade commented on app/screens/global_drafts/components/global_scheduled_post_list/global_scheduled_post_list.tsx:
```diff
@@ -0,0 +1,180 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
```
https://mattermost.atlassian.net/browse/MM-62286

Rajat-Dabade commented on app/screens/global_drafts/components/global_scheduled_post_list/index.tsx:
```diff
@@ -0,0 +1,33 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
```
Added to the list: [mattermost.atlassian.net/browse/MM-62286](https://mattermost.atlassian.net/browse/MM-62286)

Rajat-Dabade commented on app/screens/global_drafts/components/tabbed_contents/draftTabsHeader.test.tsx:
```diff
@@ -0,0 +1,69 @@
+    it('renders correctly with default props', async () => {
+        renderWithIntlAndTheme(<DraftTabsHeader {...baseProps}/>);
+
+        await act(async () => {
+            await TestHelper.wait(200); // Wait until the badge renders
+        });
```
@enahum, Can you please check this one?

Rajat-Dabade commented on app/screens/index.tsx:
```diff
@@ -114,8 +114,8 @@ Navigation.setLazyComponentRegistrator((screenName) => {
             break;
-        case Screens.DRAFT_OPTIONS:
-            screen = withServerDatabase(require('@screens/draft_options').default);
+        case Screens.DRAFT_SCHEDULED_POST_OPTIONS:
```
This file doesn’t have any tests yet, so we need to add them. I’m putting together a list of files used in this PR that don’t have tests initially written (all the reused files), and once I have the full list, I’ll create a JIRA ticket for it. We can take this up in a separate PR.

Rajat-Dabade commented on app/screens/global_drafts/components/tabbed_contents/tabbed_contents.test.tsx:
```diff
@@ -0,0 +1,131 @@
+        renderWithIntlAndTheme(<TabbedContents {...defaultProps}/>);
+
+        await act(async () => {
+            await TestHelper.wait(300); // Wait until the badge renders
```
@enahum ⬆️

Rajat-Dabade commented on app/screens/global_drafts/index.tsx:
```diff
@@ -89,13 +117,33 @@ const GlobalDrafts = ({componentId}: Props) => {
 };
 
-export default GlobalDrafts;
+const enhanced = withObservables([], ({database}: WithDatabaseArgs) => {
```
Done.

Rajat-Dabade commented on app/screens/global_drafts/index.tsx:
```diff
@@ -89,13 +117,33 @@ const GlobalDrafts = ({componentId}: Props) => {
+                        <TabbedContents
+                            initialTab={initialTab || DRAFT_SCREEN_TAB_DRAFTS}
+                            draftsCount={draftsCount}
+                            scheduledPostCount={scheduledPostCount}
```
Yup, that's true. TabbedContents is responsible for fetching the count. Updated the code.

Rajat-Dabade commented on app/screens/global_drafts/global_drafts.test.tsx:
```diff
@@ -0,0 +1,109 @@
+        );
+
+        await act(async () => {
+            await TestHelper.wait(200); // Wait until the badge renders
```
I think, this is added to ensure the draft_tab renders fully before accessing it, preventing test failures due to asynchronous updates. 0/5. @enahum ⬆️

Rajat-Dabade commented on app/screens/home/channel_list/categories_list/categories_list.tsx:
```diff
@@ -96,23 +102,25 @@ const CategoriesList = ({
 
     const draftsButtonComponent = useMemo(() => {
-        if (draftsCount > 0) {
+        if (draftsCount > 0 || scheduledPostCount > 0 || (isTablet && activeScreen === DRAFT)) {
```
Yup, that can happen. Nice Catch 🙌 . I added the check and tests.

Rajat-Dabade commented on app/screens/reschedule_draft/index.tsx:
```diff
@@ -0,0 +1,21 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
```
https://mattermost.atlassian.net/browse/MM-62286, can be taken up later.

Rajat-Dabade commented on app/screens/reschedule_draft/reschedule_draft.test.tsx:
```diff
@@ -0,0 +1,275 @@
+        expect(saveButton?.enabled).toBeTruthy();
+    });
+
+    it('should call Navigation event when save button is pressed', async () => {
```
Yes, I initially included some extra checks in the first one. I’ve removed those and kept only the check that matches the name's intent.

Rajat-Dabade commented on app/screens/reschedule_draft/reschedule_draft.test.tsx:
```diff
@@ -0,0 +1,275 @@
+            });
+        });
+
+        expect(dismissModal).toHaveBeenCalledWith({componentId: baseProps.componentId});
```
No, the close button should also dismiss the modal, we already have a test for it with name `should dismiss modal when close button is pressed`

Rajat-Dabade commented on app/screens/scheduled_post_options/core_options/index.ts:
```diff
@@ -0,0 +1,20 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
```
https://mattermost.atlassian.net/browse/MM-62286, can be taken up later.

Rajat-Dabade commented on app/screens/scheduled_post_options/core_options/core_options.tsx:
```diff
@@ -0,0 +1,157 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
```
Added to the list, can be taken up later.

Rajat-Dabade commented on app/screens/scheduled_post_options/footer/scheduled_post_footer.tsx:
```diff
@@ -0,0 +1,136 @@
+                paddingBottom: Platform.select({ios: (isTablet ? FOOTER_PADDING_BOTTOM_TABLET_ADJUST : 0), default: FOOTER_PADDING}),
+            }]}
+        >
+            <Button
```
@harshilsharma63 ⬆️

Rajat-Dabade commented on app/screens/scheduled_post_options/footer/scheduled_post_footer.test.tsx:
```diff
@@ -0,0 +1,139 @@
+    });
+
+    it('renders correctly for tablet', () => {
+        (useIsTablet as jest.Mock).mockReturnValueOnce(true);
```
Done.

Rajat-Dabade commented on app/screens/snack_bar/index.tsx:
```diff
@@ -270,12 +305,10 @@ const SnackBar = ({
+                            message={message}
                             style={toastStyle}
                             textStyle={styles.text}
+                            testID='toast'
```
The component name says `Toast`, I think the testID should be `toast`.

Rajat-Dabade commented on app/utils/scheduled_post/index.ts:
```diff
@@ -0,0 +1,128 @@
+export const hasScheduledPostError = (scheduledPosts: ScheduledPostModel[]) =>
+    scheduledPosts.some((post) => post.errorCode !== '');
+
+export type ScheduledPostErrorCode = 'unknown' | 'channel_archived' | 'channel_not_found' | 'user_missing' | 'user_deleted' | 'no_channel_permission' | 'no_channel_member' | 'thread_deleted' | 'unable_to_send' | 'invalid_post' | 'post_send_success_delete_failed';
```
Yup, moved it to the types folder.

Rajat-Dabade commented on app/utils/scheduled_post/scheduled_post.test.ts:
```diff
@@ -0,0 +1,316 @@
+    });
+});
+
+describe('getErrorStringFromCode', () => {
```
Yup, repeated. Removed one.

Rajat-Dabade commented:
Hey @larkox , I haven’t addressed some of the review comments related to tests yet, but all the blocking changes have been taken care of. I’ve created an issue to revisit the test-related changes after merging this PR: https://mattermost.atlassian.net/browse/MM-63631. We can tackle those later. Thank you for reviewing 🙌 .

Rajat-Dabade commented:
/update-branch

asaadmahmood reviewed (changes requested):
I'm seeing a bunch of bugs.

https://github.com/user-attachments/assets/4f12273b-1ab1-4f00-9061-c104b7f464af

Also some updates needed.
![IMG_0929](https://github.com/user-attachments/assets/665e392a-d033-4f71-a0f3-eba714809c29)

![CleanShot 2025-04-04 at 17 00 58@2x](https://github.com/user-attachments/assets/51863609-71e1-422a-b775-7e50c22a55d7)

asaadmahmood commented:
![CleanShot 2025-04-04 at 17 05 32@2x](https://github.com/user-attachments/assets/76fdd594-25ef-4ffe-9fc9-95a71ea1da5d)

enahum commented:
My two cents based on the comment of fitting the error label by @asaadmahmood 

You will need to take into consideration that this text is translated to many other languages and detecting if it fits or not may not be the best thing to do here.

I would advice for you guys to work together on a solution that looks well and applies to all languages

Rajat-Dabade commented:
/update-branch

mattermost-build commented:
Error trying to update the PR.
Please do it manually.

larkox commented on app/screens/draft_scheduled_post_options/send_draft/index.ts:
```diff
@@ -0,0 +1,71 @@
+            }
+            const teammateId = getUserIdFromChannelName(u.id, c.name);
+            if (teammateId) {
+                return observeUser(database, teammateId).pipe(
+                    switchMap((u2) => (u2 ? of$(Boolean(u2.deleteAt)) : of$(false))), // eslint-disable-line max-nested-callbacks
+                );
```
In order to avoid such deep nesting, you can pass a another argument to the pipe. Something like...

```
const deactivatedChannel = combineLatest([currentUser, channel]).pipe(
    switchMap(([u, c]) => {
        ....
        return observeUser(database, temamateId);
    }),
    switchMap((u2) => (u2 ? of$(Boolean(u2.deleteAt)) : of$(false)))
);
``

Rajat-Dabade replied:
Done.

larkox commented on app/screens/global_drafts/components/tabbed_contents/index.ts:
```diff
@@ -0,0 +1,25 @@
+
+const enhanced = withObservables([], ({database}: WithDatabaseArgs) => {
+    const currentTeamId = observeCurrentTeamId(database);
+    const draftsCount = currentTeamId.pipe(switchMap((teamId) => observeDraftCount(database, teamId))); // Observe draft count
```
Unneeded comment.

larkox reviewed (approved):
A couple of minor comments, but LGTM

yasserfaraazkhan reviewed (approved):
Thank you Rajat,
Verified the final changes on android and iOS device. Log few no blocking defect in the epic.

enahum commented on app/screens/scheduled_post_options/scheduled_post_picker.tsx:
```diff
@@ -0,0 +1,148 @@
+            renderContent={renderContent}
+            componentId={Screens.SCHEDULED_POST_OPTIONS}
+            closeButtonId={SCHEDULED_POST_OPTIONS_BUTTON}
+            snapPoints={snapPoints}
```
For this one, it seems that there are times where no options show on screen, this will leave an empty space in the bottom sheet initial snap point, I think you should consider using the dynamic sizing provided by the library for this use case.

Rajat-Dabade replied:
This can be taken up later, we already have a ticket about bottom sheet and fixing up the height. I will put the link of the conversion there to track.

Rajat-Dabade replied:
https://mattermost.atlassian.net/browse/MM-63710

enahum commented on app/screens/scheduled_post_options/scheduled_post_picker.tsx:
```diff
@@ -0,0 +1,148 @@
+    const style = getStyleSheet(theme);
+
+    const snapPoints = useMemo(() => {
+        const bottomSheetAdjust = Platform.select({ios: 5, default: 20});
```
I don't like this, this should probably use the bottom inset

Rajat-Dabade replied:
This can also be taken up later. https://mattermost.atlassian.net/browse/MM-63710

enahum commented on app/screens/scheduled_post_options/scheduled_post_picker.tsx:
```diff
@@ -0,0 +1,148 @@
+        const bottomSheetAdjust = Platform.select({ios: 5, default: 20});
+
+        // 9 items to display inline date-time picker, 4 items otherwise
+        const iosNumberOfItems = customTimeSelected ? 9 : 4;
+        const andriodNumberOfItems = customTimeSelected ? 4 : 3;
+        const numberOfItems = Platform.select({ios: iosNumberOfItems, default: andriodNumberOfItems});
```
I don;'t like this either, probably best to use dynamic sizing

Rajat-Dabade replied:
The same can be taken up later.

enahum commented on app/screens/scheduled_post_options/scheduled_post_picker.tsx:
```diff
@@ -0,0 +1,148 @@
+    }, []);
+
+    const handleOnSchedule = usePreventDoubleTap(useCallback(async () => {
+        if (!selectedTime) {
```
is there a reason to log this but not to display anything to the user ?

Rajat-Dabade replied:
Added the isError state to show the error of not selecting time.

enahum commented on app/screens/scheduled_post_options/scheduled_post_picker.tsx:
```diff
@@ -0,0 +1,148 @@
+            const errorMessage = response.error as string;
+            showScheduledPostCreationErrorSnackbar(errorMessage);
+        } else {
+            await dismissBottomSheet();
```
what is the point of the await here?

Rajat-Dabade replied:
No need.

enahum commented on app/screens/scheduled_post_options/scheduled_post_picker.tsx:
```diff
@@ -0,0 +1,148 @@
+        const response = await onSchedule(schedulingInfo);
+        setIsScheduling(false);
+
+        if (response?.error) {
+            const errorMessage = response.error as string;
+            showScheduledPostCreationErrorSnackbar(errorMessage);
+        } else {
+            await dismissBottomSheet();
+        }
```
```suggestion
        if (response?.error) {
            const errorMessage = response.error as string;
            showScheduledPostCreationErrorSnackbar(errorMessage);
            return;
        }
        dismissBottomSheet();
```

enahum reviewed (changes requested):
have not finished the review, but some comments so that you can address them in the meantime

Rajat-Dabade commented on app/actions/local/scheduled_post.ts:
```diff
@@ -0,0 +1,34 @@
+        const models = await operator.handleScheduledPosts({actionType, scheduledPosts, prepareRecordsOnly});
+        return {models};
+    } catch (error) {
+        logError('handleScheduledPosts cannot handle scheduled post websocket event', error);
```
It won’t log the error twice because we’re catching and logging the error within each function without re-throwing it. The error is handled at the point it occurs, so it only gets logged once.

Rajat-Dabade replied:
Removed the websocket part.

Rajat-Dabade commented on app/actions/local/scheduled_post.ts:
```diff
@@ -0,0 +1,34 @@
+
+export async function handleScheduledPosts(serverUrl: string, actionType: string, scheduledPosts: ScheduledPost[], prepareRecordsOnly = false): Promise<{models?: ScheduledPostModel[]; error?: unknown}> {
+    if (!scheduledPosts.length) {
+        return {models: undefined};
```
Referenced it from the ChannelBookmark code. We can change if that's the preferred way.

enahum replied:
maybe I made a mistake in bookmarks, but I kind of leaning to the empty array, anyway we can keep this as is

Rajat-Dabade commented on app/actions/remote/scheduled_post.ts:
```diff
@@ -0,0 +1,127 @@
+    }
+}
+
+export async function updateScheduledPost(serverUrl: string, scheduledPost: ScheduledPost | ScheduledPostModel, updateTime: number, connectionId?: string, fetchOnly = false) {
```
Ah yes, we can do the same here. Will update it.

Rajat-Dabade commented on app/actions/local/scheduled_post.ts:
```diff
@@ -0,0 +1,34 @@
+
+import type {ScheduledPostErrorCode} from '@typings/utils/scheduled_post';
+
+export async function handleScheduledPosts(serverUrl: string, actionType: string, scheduledPosts: ScheduledPost[], prepareRecordsOnly = false): Promise<{models?: ScheduledPostModel[]; error?: unknown}> {
```
Changed it.

Rajat-Dabade commented on app/client/rest/scheduled_post.ts:
```diff
@@ -0,0 +1,57 @@
+            {
+                method: 'post',
+                body: schedulePost,
+                headers: {'Connection-Id': connectionId},
```
Good point! You're right, it's better to avoid including the Connection-Id header if the connectionId is absent. While sending it as an empty string might not break anything, it's cleaner and avoids unnecessary headers in the request. I have taken the reference from the channelBookmark code, so we should probably make changes there as well in the future.

Rajat-Dabade commented on app/components/draft_scheduled_post/draft_scheduled_post.tsx:
```diff
@@ -0,0 +1,155 @@
+import {Screens} from '@constants';
+import {useServerUrl} from '@context/server';
+import {useTheme} from '@context/theme';
+import {DRAFT_OPTIONS_BUTTON} from '@screens/draft_scheduled_post_options';
+import {DRAFT_TYPE_DRAFT, DRAFT_TYPE_SCHEDULED, type DraftType} from '@screens/global_drafts/constants';
```
Moved it.

Rajat-Dabade commented on app/components/draft_scheduled_post/draft_scheduled_post.tsx:
```diff
@@ -0,0 +1,155 @@
+
+import DraftAndScheduledPostContainer from './draft_scheduled_post_container';
+
+import type {ScheduledPostModel} from '@database/models/server';
```
Nice catch, will update it.

Rajat-Dabade commented on app/components/post_draft/send_handler/send_handler.tsx:
```diff
@@ -4,13 +4,13 @@
-import DraftInput from '../draft_input';
+import SendDraft from '@screens/draft_scheduled_post_options/send_draft';
 
+import type {DraftType} from '@screens/global_drafts/constants';
```
Moved it.

Rajat-Dabade commented on app/components/scheduled_post_indicator/index.ts:
```diff
@@ -0,0 +1,35 @@
+    rootId?: string;
+}
+
+const enhance = withObservables([], ({database, channelId, rootId}: Props) => {
```
Yes, true.

Rajat-Dabade commented on app/components/scheduled_post_indicator/scheduled_post_indicator.tsx:
```diff
@@ -0,0 +1,124 @@
+        },
+        text: {
+            color: changeOpacity(theme.centerChannelColor, 0.75),
+            fontSize: 14,
```
Yup, updated it.

enahum commented on app/components/draft_scheduled_post_header/draft_scheduled_post_header.tsx:
```diff
@@ -64,23 +75,54 @@ const getStyleSheet = makeStyleSheetFromTheme((theme: Theme) => {
             ...typography('Body', 75),
         },
+        scheduledContainer: {
+            display: 'flex',
```
I see a bunch of these everywhere, this is not needed as is the default display on RN, ok to keep but just mentioning that is not needed.

Rajat-Dabade replied:
REmoved it.

enahum commented on app/screens/reschedule_draft/reschedule_draft.tsx:
```diff
@@ -0,0 +1,151 @@
+    draft: ScheduledPostModel;
+}
+
+const OPTIONS_PADDING = 12;
```
this is only used in the style and just once, do we really need to define the constant?

Rajat-Dabade replied:
Done.

enahum commented on app/screens/scheduled_post_options/core_options/core_options.tsx:
```diff
@@ -0,0 +1,157 @@
+const optionKeyOptionNextMonday = 'scheduledPostOptionNextMonday';
+const optionKeyOptionCustom = 'scheduledPostOptionCustom';
+
+const OPTIONS_SEPARATOR_HEIGHT = 1;
```
same as the previous file

Rajat-Dabade replied:
Done.

enahum commented on app/screens/scheduled_post_options/footer/scheduled_post_footer.tsx:
this file can be rename as the index file

Rajat-Dabade replied:
Done.

enahum commented on app/screens/scheduled_post_options/footer/scheduled_post_footer.tsx:
```diff
@@ -0,0 +1,136 @@
+        ...typography('Body', 200, 'SemiBold'),
+    },
+    applyButtonProcessingText: {
+        color: changeOpacity(theme.centerChannelColor, 0.32),
```
no typography applied to this one ?

Rajat-Dabade replied:
This should be the same as the one above. Updated it.

enahum reviewed (changes requested):
still pending changes + new ones

Rajat-Dabade commented on app/components/post_draft/send_handler/send_handler.tsx:
```diff
@@ -4,13 +4,13 @@
-import SendDraft from '@screens/draft_options/send_draft';
-
-import DraftInput from '../draft_input';
+import SendDraft from '@screens/draft_scheduled_post_options/send_draft';
```
Moved it to `@components`

Rajat-Dabade commented on app/components/scheduled_post_indicator/scheduled_post_indicator.tsx:
```diff
@@ -0,0 +1,124 @@
+        return null;
+    }
+    let scheduledPostText;
+    if (scheduledPostCount === 1) {
+        scheduledPostText = isThread ? (
+            <FormattedMessage
+                id='scheduled_post.channel_indicator.thread.single'
+                defaultMessage='1 scheduled message in thread.'
+            />
+        ) : (
+            <FormattedMessage
+                id='scheduled_post.channel_indicator.single'
+                defaultMessage='1 scheduled message in channel.'
+            />
+        );
+    } else {
+        scheduledPostText = isThread ? (
+            <FormattedMessage
+                id='scheduled_post.channel_indicator.thread.multiple'
+                defaultMessage='{count} scheduled messages in thread.'
+                values={{
+                    count: scheduledPostCount,
+                }}
+            />
+        ) : (
+            <FormattedMessage
+                id='scheduled_post.channel_indicator.multiple'
+                defaultMessage='{count} scheduled messages in channel.'
+                values={{
+                    count: scheduledPostCount,
+                }}
+            />
+        );
+    }
```
This is awesome; I was not aware of it.

Rajat-Dabade commented on app/constants/tooltip.ts:
```diff
@@ -0,0 +1,9 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+export const tooltipContentStyle = {
```
Okay done.

Rajat-Dabade commented on app/database/models/server/scheduled_post.ts:
```diff
@@ -0,0 +1,82 @@
+            processed_at: this.processedAt,
+            error_code: this.errorCode,
+            create_at: this.createAt,
+            user_id: await getCurrentUserId(serverDatabase),
```
Yup, this can be done.

Rajat-Dabade commented on app/database/operator/server_data_operator/transformers/post.ts:
```diff
@@ -160,3 +162,39 @@ export const transformPostsInChannelRecord = ({action, database, value}: Transfo
+ */
+export const transformSchedulePostsRecord = ({action, database, value}: TransformerArgs<ScheduledPostModel, ScheduledPost>): Promise<ScheduledPostModel> => {
+    const emptyFileInfo: FileInfo[] = [];
+    const raw = value.raw;
```
Sure, updated it. And also added a test for the same.

Rajat-Dabade commented on app/database/schema/server/index.ts:
```diff
@@ -40,7 +41,7 @@ import {
 
 export const serverSchema: AppSchema = appSchema({
-    version: 7,
+    version: 8,
```
Sure.

Rajat-Dabade commented on app/database/schema/server/table_schemas/scheduled_post.ts:
We can take this up in another PR.

Rajat-Dabade replied:
https://mattermost.atlassian.net/browse/MM-63758

Rajat-Dabade commented on app/database/schema/server/test.ts:
```diff
@@ -46,7 +47,7 @@ const {
     it('=> The SERVER SCHEMA should strictly match', () => {
         expect(serverSchema).toEqual({
-            version: 7,
+            version: 8,
```
Resolved it.

Rajat-Dabade commented on app/hooks/handle_send_message.ts:
```diff
@@ -167,31 +222,34 @@ export const useHandleSendMessage = ({
         const toHere = DraftUtils.textContainsAtHere(value);
 
-        if (value.indexOf('/') === 0) {
+        if (value.indexOf('/') === 0 && !schedulingInfo) {
```
No slash command is not a scheduled post

Rajat-Dabade replied:
This check is currently needed to maintain consistency with the webapp’s behavior when handling messages that start with a slash.

harshilsharma63 replied:
@enahum the feature allows you to schedule a slash command, but it won't be executed. For example, if you type `/command` and schedule it, it will post the message with text `/command`. Hence, the check for schedulingInfo. This is as per specs and as per what happens in webapp. I believe you misunderstood this as a slash command should disable the scheduling option, which is not whats desired.

Rajat-Dabade commented on types/database/models/servers/scheduled_post.ts:
```diff
@@ -0,0 +1,49 @@
+    /** error_code : The reason message if the schedule post failed */
+    errorCode: string;
+
+    toApi: (serverDatabase: Database) => Promise<ScheduledPost>;
```
Yup, done.

Rajat-Dabade commented on types/api/config.d.ts:
```diff
@@ -200,6 +201,7 @@ interface ClientConfig {
     WebsocketPort: string;
     WebsocketSecurePort: string;
     WebsocketURL: string;
+    ScheduledPosts: string;
```
Yes, Weird. Removed the extra one.

Rajat-Dabade commented on app/utils/snack_bar/index.ts:
```diff
@@ -54,3 +63,12 @@ export const showThreadFollowingSnackbar = (following: boolean, onAction: () =>
+    return showSnackBar({
+        barType: SNACK_BAR_TYPE.SCHEDULED_POST_CREATION_ERROR,
+        customMessage: errorMessage,
+        keepOpen: true,
```
Ah, this was missed. We are not keeping it open. Will remove it.

Rajat-Dabade commented on app/screens/snack_bar/index.tsx:
```diff
@@ -3,7 +3,14 @@
+    DeviceEventEmitter,
+    Text,
+    TouchableOpacity,
+    useWindowDimensions,
```
Okay.

harshilsharma63 commented on app/screens/snack_bar/index.tsx:
```diff
@@ -86,6 +96,9 @@ const SnackBar = ({
     componentId,
     onAction,
     sourceScreen,
+    keepOpen,
```
It needs to stay open to allow the user enough time to read the error message. Otherwise it closes in 3 seconds, which is pretty useless if you're trying to figure out what went wrong, and since there is no way to see the message back again. its fine if it stays open even if you go to other screen. Its better than showing an error message and making is unable to read for the user. Also consider that its open set true for create scheduled post feature, so it doesn't hurt anything else. . 

We can tweak this, or remove this if there is a strong objection but unless its a blocker, I'd like to keep it for the sake of delivering the feature.

harshilsharma63 replied:
Okay, for the sake of this comment, I'm removing the keep open param. Its more important for me that the feature gets available to the users first, and my concern with the timing on the toast can be addressed later.

harshilsharma63 commented on app/hooks/handle_send_message.ts:
```diff
@@ -167,31 +222,34 @@ export const useHandleSendMessage = ({
+            showSendToAllOrChannelOrHereAlert(membersCount, toHere && !toAllOrChannel, schedulingInfo);
         } else {
-            doSubmitMessage();
+            return doSubmitMessage(schedulingInfo);
```
See here - https://github.com/mattermost/mattermost-mobile/pull/8509/files#diff-9719c6743ab40e564682948af82474ba889aa50ac12c66b51915dc9cc427e69cR95

Its needed because I need the response by the function to get the error message, if any, that occurred during creating the scheduled post.

Rajat-Dabade commented on app/components/post_draft/send_handler/index.ts:
```diff
@@ -20,19 +22,15 @@ import type {WithDatabaseArgs} from '@typings/database/database';
-    } else {
-        channel = observeCurrentChannel(database);
-    }
+    const {rootId, channelId, draftType} = ownProps;
```
Okay

Rajat-Dabade commented on app/queries/servers/scheduled_post.ts:
```diff
@@ -0,0 +1,68 @@
+
+export const queryScheduledPost = (database: Database, channelId: string, rootId = '') => {
+    return database.collections.get<ScheduledPostModel>(SCHEDULED_POST).query(
+        Q.where('channel_id', channelId),
+        Q.where('root_id', rootId),
```
If we don't mention anything and put 2 or more where conditions, it's by default `AND` condition, but will add `Q.and` for more readability.

Rajat-Dabade commented on app/queries/servers/scheduled_post.ts:
```diff
@@ -0,0 +1,68 @@
+    isCRTEnabled: boolean,
+) => {
+    let query = database.get<ScheduledPostModel>(SCHEDULED_POST).query(
+        Q.on(CHANNEL,
+            Q.where('id', channelId),
+        ),
```
Yup, removed it.

Rajat-Dabade commented on app/screens/draft_scheduled_post_options/index.tsx:
```diff
@@ -7,22 +7,27 @@ import {Platform, View} from 'react-native';
 import {useTheme} from '@context/theme';
 import {useIsTablet} from '@hooks/device';
 import BottomSheet from '@screens/bottom_sheet';
+import CopyTextOption from '@screens/post_options/options/copy_text_option';
```
Okay

Rajat-Dabade commented on app/screens/draft_scheduled_post_options/index.tsx:
```diff
@@ -53,26 +59,43 @@ const DraftOptions: React.FC<Props> = ({
         const bottomSheetAdjust = Platform.select({ios: 5, default: 20});
-        const COMPONENT_HIEGHT = TITLE_HEIGHT + (3 * ITEM_HEIGHT) + bottomSheetAdjust;
-        return [1, COMPONENT_HIEGHT];
+        const COMPONENT_HEIGHT = TITLE_HEIGHT + (4 * ITEM_HEIGHT) + bottomSheetAdjust;
```
Taken the reference from `app/screens/post_priority_picker/post_priority_picker.tsx`

Rajat-Dabade replied:
Updated the code.

Rajat-Dabade commented on app/screens/global_drafts/components/global_scheduled_post_list/global_scheduled_post_list.tsx:
```diff
@@ -0,0 +1,189 @@
+
+    useAndroidHardwareBackHandler(Screens.GLOBAL_DRAFTS, collapse);
+
+    const close = useCallback(() => {
```
Done.

Rajat-Dabade commented on app/screens/global_drafts/components/global_scheduled_post_list/index.tsx:
```diff
@@ -0,0 +1,33 @@
+    teamId: string;
+} & WithDatabaseArgs;
+
+const withTeamId = withObservables([], ({database}: WithDatabaseArgs) => ({
+    teamId: observeCurrentTeamId(database),
+}));
+
+const enhanced = withObservables(['teamId'], ({database, teamId}: Props) => {
+    const allScheduledPosts = observeScheduledPostsForTeam(database, teamId, true);
+    const tutorialWatched = observeTutorialWatched(Tutorial.SCHEDULED_POSTS_LIST);
+
+    return {
+        allScheduledPosts,
+        tutorialWatched,
+    };
+});
```
Okay done.

Rajat-Dabade commented on app/screens/global_drafts/components/tabbed_contents/draftTabsHeader.tsx:
```diff
@@ -0,0 +1,174 @@
+    const styles = getStyleSheet(theme);
+    const viewingDrafts = selectedTab === DRAFT_TAB_INDEX;
+
+    const draftCountBadge = useMemo(() => {
+        const style = [styles.badgeStyles, selectedTab === DRAFT_TAB_INDEX ? styles.activeBadgeStyles : null];
+
+        return (
+            <Badge
+                value={draftsCount}
+                visible={draftsCount !== 0}
+                style={style}
+                testID='draft_count_badge'
+            />
+        );
+    }, [draftsCount, selectedTab, styles.activeBadgeStyles, styles.badgeStyles]);
+
+    const scheduledPostCountBadge = useMemo(() => {
+        const style = [styles.badgeStyles, selectedTab === DRAFT_SCREEN_TAB_SCHEDULED_POSTS ? styles.activeBadgeStyles : null];
+
+        return (
+            <Badge
+                value={scheduledPostCount}
+                visible={scheduledPostCount !== 0}
+                style={style}
+                testID='scheduled_post_count_badge'
+            />
+        );
+    }, [scheduledPostCount, selectedTab, styles.activeBadgeStyles, styles.badgeStyles]);
```
Yes, you are right; there is no value in memorizing these. Nice catch. Will update it.

Rajat-Dabade commented on app/actions/remote/team.ts:
```diff
@@ -436,6 +437,7 @@ export async function handleTeamChange(serverUrl: string, teamId: string) {
 
     // Fetch Groups + GroupTeams
     fetchGroupsForTeamIfConstrained(serverUrl, teamId);
+    fetchScheduledPosts(serverUrl, teamId, false);
```
We can take this up later. Create a ticket to track it: https://mattermost.atlassian.net/browse/MM-63737

Rajat-Dabade replied:
Later is before the code freeze (before the release of this feature), as I am on SET right now, not able to give much time to this.

Rajat-Dabade commented on app/components/draft_scheduled_post_header/draft_scheduled_post_header.tsx:
```diff
@@ -135,25 +177,48 @@ const DraftPostHeader: React.FC<Props> = ({
-                <Text style={style.displayName}>
-                    {channel.displayName}
-                </Text>
+        <View>
+            {draftType === DRAFT_TYPE_SCHEDULED &&
+                <View style={style.scheduledContainer}>
+                    <Text style={style.scheduledAtText}>
+                        {scheduledPostErrorCode === 'post_send_success_delete_failed' ? intl.formatMessage({id: 'scheduled_post.header.sent', defaultMessage: 'Sent'}) : intl.formatMessage({id: 'channel_info.scheduled', defaultMessage: 'Send on {time}'}, {time: getReadableTimestamp(postScheduledAt!, getUserTimezone(currentUser), isMilitaryTime, currentUser?.locale || DEFAULT_LOCALE)})}
+                    </Text>
+                    {scheduledPostErrorCode &&
+                        <View style={style.errorState}>
+                            <CompassIcon
+                                name='alert-outline'
+                                size={12}
+                                color={theme.buttonColor}
+                            />
+                            <Text style={style.errorText}>{getErrorStringFromCode(intl, scheduledPostErrorCode as ScheduledPostErrorCode)}</Text>
```
Done.

Rajat-Dabade commented on ios/Gekidou/Package.resolved:
Okay.

Rajat-Dabade commented on app/database/migration/server/index.ts:
```diff
@@ -8,9 +8,29 @@ import {addColumns, createTable, schemaMigrations, unsafeExecuteSql} from '@nozb
 
 export default schemaMigrations({migrations: [
+    {
+        toVersion: 8,
```
Resolved it.

harshilsharma63 commented on app/screens/custom_status_clear_after/components/date_time_selector.tsx:
```diff
@@ -23,19 +23,18 @@ type Props = {
 const getStyleSheet = makeStyleSheetFromTheme((theme: Theme) => {
     return {
         container: {
-            flex: 1,
```
Looks just fine

https://github.com/user-attachments/assets/f358ee9b-d7f9-4d9f-b186-99f43fa634b0

harshilsharma63 replied:
Android-

https://github.com/user-attachments/assets/804a1963-d72c-4b7c-bd49-be83e80e8b65

yasserfaraazkhan commented:
/update-branch

Rajat-Dabade commented on app/hooks/handle_send_message.ts:
```diff
@@ -105,20 +125,55 @@ export const useHandleSendMessage = ({
+            return response;
+        }
+
+        if (shouldClearDraft) {
+            clearDraft();
+        }
```
Done.

Rajat-Dabade commented on app/components/copy_text_option/copy_text_option.tsx:
Done.

Rajat-Dabade commented on app/screens/custom_status_clear_after/components/clear_after_menu_item.tsx:
```diff
@@ -15,7 +15,7 @@ import {preventDoubleTap} from '@utils/tap';
 import {getTimezone} from '@utils/user';
 
-import DateTimePicker from './date_time_selector';
+import DateTimePicker from '../../../components/data_time_selector/date_time_selector';
```
Done.

Rajat-Dabade commented on app/components/scheduled_post_indicator/scheduled_post_indicator.tsx:
Done.

Rajat-Dabade commented on app/screens/draft_scheduled_post_options/index.tsx:
```diff
@@ -0,0 +1,146 @@
+import {makeStyleSheetFromTheme} from '@utils/theme';
+import {typography} from '@utils/typography';
+
+import DeleteDraft from '../../components/draft_scheduled_post/draft_scheduled_post_actions/delete_draft';
+import EditDraft from '../../components/draft_scheduled_post/draft_scheduled_post_actions/edit_draft';
+import RescheduledDraft from '../../components/draft_scheduled_post/draft_scheduled_post_actions/rescheduled_draft';
```
Done.

Rajat-Dabade commented on app/screens/draft_scheduled_post_options/index.tsx:
```diff
@@ -0,0 +1,146 @@
+                    cursorPosition={0}
+                    draftType={draftType}
+                    postId={draft.id}
+                    /* eslint-disable no-empty-function */
+                    clearDraft={() => {}}
+                    updateCursorPosition={() => {}}
+                    updatePostInputTop={() => {}}
+                    addFiles={() => {}}
+                    setIsFocused={() => {}}
+                    updateValue={() => {}}
+                    /* eslint-enable no-empty-function */
```
Done.

Rajat-Dabade commented on app/screens/global_drafts/components/tabbed_contents/draftTabsHeader.tsx:
Done.

harshilsharma63 commented on app/components/data_time_selector/date_time_selector.tsx:
```diff
@@ -23,19 +23,18 @@ type Props = {
 const getStyleSheet = makeStyleSheetFromTheme((theme: Theme) => {
     return {
         container: {
-            flex: 1,
```
I believe I previously replied to the same comment but it appears to be unavailable now. The change doesn't negatively the only other place - custom date time picker for custom status, as was checked during development. See attached videos-

https://github.com/user-attachments/assets/37ac1b88-2188-4ec7-863e-408dabdd6c00

https://github.com/user-attachments/assets/5a4970f8-ee27-45f8-b21e-12d9de8e97cf

harshilsharma63 replied:
@enahum here's your previous comment asking the same thing and me replying the same thing there - https://github.com/mattermost/mattermost-mobile/pull/8509#discussion_r2036305457

enahum commented on app/screens/post_options/post_options.tsx:
```diff
@@ -18,8 +18,9 @@ import {dismissBottomSheet} from '@screens/navigation';
 import {bottomSheetSnapPoint} from '@utils/helpers';
 import {isSystemMessage} from '@utils/post';
 
+import CopyTextOption from '../../components/copy_text_option';
```
fix import

Rajat-Dabade replied:
Done.

enahum reviewed (approved):
one small nit but approving

yasserfaraazkhan reviewed (approved):
Thanks Rajat,
I was able to verify the scenarios. there are few UI issues in android. I've logged them under the epic.

Rajat-Dabade commented:
Security review can be done before the code freeze for this feature. 
cc// @esarafianou

amyblais commented:
/cherry-pick release-2.28

mattermost-build commented:
Cherry pick is scheduled.

esarafianou reviewed (commented):
LGTM

cwarnermm commented:
@Rajat-Dabade - On mobile, is it a long press on the Send button that allows users to schedule their message to send at a future time?

Rajat-Dabade commented:
> On mobile, is it a long press on the Send button that allows users to schedule their message to send at a future time?

@cwarnermm Yes, it is.
